File: src/main/java/club/sk1er/mods/core/universal/UniversalPacket.java
package club.sk1er.mods.core.universal;

import club.sk1er.mods.core.universal.wrappers.message.UniversalTextComponent;

//#if MC>=11602
//$$ import club.sk1er.mods.core.universal.wrappers.UniversalPlayer;
//#endif

//#if MC>=11502
//$$ import net.minecraft.network.play.server.SChatPacket;
//$$ import net.minecraft.util.text.ChatType;
//#elseif MC>=11202
//$$ import net.minecraft.network.play.server.SPacketChat;
//$$ import net.minecraft.util.text.ChatType;
//#else
import net.minecraft.network.play.server.S02PacketChat;
//#endif

public class UniversalPacket {
    public static void sendChatMessage(UniversalTextComponent component) {
        //#if FABRIC
        //$$ UniversalPlayer.getPlayer().sendChatMessage(component.getText());
        //#elseif MC>=11602
        //$$ UniversalMinecraft.getNetHandler().handleChat(new SChatPacket(component, ChatType.CHAT, UniversalPlayer.getUUID()));
        //#elseif MC>=11502
        //$$ UniversalMinecraft.getNetHandler().handleChat(new SChatPacket(component, ChatType.CHAT));
        //#elseif MC>=11202
        //$$ UniversalMinecraft.getNetHandler().handleChat(new SPacketChat(component, ChatType.CHAT));
        //#else
        UniversalMinecraft.getNetHandler().handleChat(new S02PacketChat(component, (byte) 0));
        //#endif
    }

    public static void sendActionBarMessage(UniversalTextComponent component) {
        //#if FABRIC
        //$$ throw new UnsupportedOperationException("TODO");
        //#elseif MC>=11602
        //$$ UniversalMinecraft.getNetHandler().handleChat(new SChatPacket(component, ChatType.GAME_INFO, UniversalPlayer.getUUID()));
        //#elseif MC>=11502
        //$$ UniversalMinecraft.getNetHandler().handleChat(new SChatPacket(component, ChatType.GAME_INFO));
        //#elseif MC>=11202
        //$$ UniversalMinecraft.getNetHandler().handleChat(new SPacketChat(component, ChatType.GAME_INFO));
        //#else
        UniversalMinecraft.getNetHandler().handleChat(new S02PacketChat(component, (byte) 2));
        //#endif
    }
}

File: src/main/kotlin/club/sk1er/mods/core/universal/UniversalPacket.kt
package club.sk1er.mods.core.universal

import club.sk1er.mods.core.universal.utils.MCSChatPacket
import club.sk1er.mods.core.universal.wrappers.message.UniversalTextComponent
import club.sk1er.mods.core.universal.wrappers.UniversalPlayer

//#if FORGE && MC>=11202
//$$ import net.minecraft.util.text.ChatType
//#endif

object UniversalPacket {
    @JvmStatic
    fun sendChatMessage(message: UniversalTextComponent) {
        //#if FABRIC
        //$$ UniversalPlayer.getPlayer()!!.sendChatMessage(message.text)
        //#elseif MC>=11602
        //$$ UniversalMinecraft.getNetHandler()!!.handleChat(MCSChatPacket(message, ChatType.CHAT, UniversalPlayer.getUUID()))
        //#elseif MC>=11502
        //$$ UniversalMinecraft.getNetHandler()!!.handleChat(MCSChatPacket(message, ChatType.CHAT))
        //#elseif MC>=11202
        //$$ UniversalMinecraft.getNetHandler()!!.handleChat(MCSChatPacket(message, ChatType.CHAT))
        //#else
        UniversalMinecraft.getNetHandler()!!.handleChat(MCSChatPacket(message, 0))
        //#endif
    }
    
    @JvmStatic
    fun sendActionBarMessage(message: UniversalTextComponent) {
        //#if FABRIC
        //$$ TODO("Figure out how to do this with Fabric")
        //#elseif MC>=11602
        //$$ UniversalMinecraft.getNetHandler()!!.handleChat(MCSChatPacket(message, ChatType.GAME_INFO, UniversalPlayer.getUUID()))
        //#elseif MC>=11502
        //$$ UniversalMinecraft.getNetHandler()!!.handleChat(MCSChatPacket(message, ChatType.GAME_INFO))
        //#elseif MC>=11202
        //$$ UniversalMinecraft.getNetHandler()!!.handleChat(MCSChatPacket(message, ChatType.GAME_INFO))
        //#else
        UniversalMinecraft.getNetHandler()!!.handleChat(MCSChatPacket(message, 2))
        //#endif
    }
}
